nomos-mix/core/src/persistent_transmission.rs

```diff
@@ -26,13 +33,45 @@ impl Default for PersistentTransmissionSettings {
 /// * `schedule_receiver` - The channel for scheduled messages
 /// * `emission_sender` - The channel to emit messages
 pub async fn persistent_transmission(
-    _settings: PersistentTransmissionSettings,
-    mut schedule_receiver: mpsc::UnboundedReceiver<Vec<u8>>,
+    settings: PersistentTransmissionSettings,
+    schedule_receiver: mpsc::UnboundedReceiver<Vec<u8>>,
     emission_sender: mpsc::UnboundedSender<Vec<u8>>,
 ) {
-    // TODO: This is a mock implementation.
-    //   The actual implementation must use emission frequency and emits drop messages.
-    while let Some(message) = schedule_receiver.recv().await {
-        emission_sender.send(message).unwrap();
+    let mut schedule_receiver = schedule_receiver;
+    let mut interval = time::interval(Duration::from_secs_f64(
+        1.0 / settings.max_emission_frequency,
+    ));
+
+    loop {
+        interval.tick().await;
+
+        // Emit the first one of the scheduled messages.
+        // If there is no scheduled message, emit a drop message with probability.
+        match schedule_receiver.try_recv() {
```

Comment:
Receiving should be independent from the interval. We cannot read messages according to ticks, we must only calculate the rate at which they are arriving but receive them instantly.

Reply:
Yes. The "schedule" channel is corresponding to the "Input Queue" in the spec. Messages arrive in the schedule channel from other modules (Tier 2, currently). 

This code (`schedule_receiver.try_recv()`) is for reading the schedule channel (Input Queue), as defined in the [spec](https://www.notion.so/Nomos-Blend-Network-Tier-1-Persistent-Transmission-Module-10b8f96fb65c807cb1e8f92a7f41a771?pvs=4#11c8f96fb65c808cb008d2595105c32b). 
> Message is read from the queue at the CEF frequency.

The term `receiver` might be little confusing to me. That's the naming convention that represents the "receiver end" of async channel in Rust.

Reply:
This should be rewrite using tokio::select probably. No need to wait for receiving as marcin says. If we received they are stored. Otherwise on interval, pop messages or send DROP ones.

Reply:
I think the current implementation is correct.

Messages are stored in the schedule channel, as soon as other modules can them to the schedule channel (using the sender half: `schedule_sender`). This channel is a mpsc unbound channel.

In this code, we sleep for an interval and "try" to "pop" one message from the schedule channel. This try is non-blocking. If the channel is empty, send a drop message.

I think the code implements the spec correctly. Please let me know if I'm missing anything. BTW, it's my bad that I didn't write tests. I'll do it today.

Reply:
494cf76dce3c21e0aa8d4b9db0a3687247284f4b I just added a test. You can see that multiple messages can be scheduled without blocking. Those messages are emitted with the expected interval. If there is no message scheduled, drop messages are emitted with the expected interval as well.